File: data/derived_cot/rq1_traces/translation/r1/py_to_cpp/code_output/VectorUtil.cpp
#include <Eigen/Dense>
#include <vector>
#include <map>
#include <cmath>
#include <stdexcept>
#include <string>

using namespace Eigen;

class VectorUtil {
public:
    static double similarity(const VectorXd& vector_1, const VectorXd& vector_2) {
        VectorXd u1 = unitvec(vector_1);
        VectorXd u2 = unitvec(vector_2);
        return u1.dot(u2);
    }

    static VectorXd cosine_similarities(const VectorXd& vector_1, const std::vector<VectorXd>& vectors_all) {
        size_t n = vectors_all.size();
        if (n == 0) {
            return VectorXd();
        }
        Index d = vector_1.size();

        MatrixXd M(n, d);
        for (size_t i = 0; i < n; ++i) {
            if (vectors_all[i].size() != d) {
                throw std::invalid_argument("All vectors in vectors_all must have the same dimension as vector_1.");
            }
            M.row(i) = vectors_all[i];
        }

        double norm1 = vector_1.norm();
        VectorXd all_norms = M.rowwise().norm();
        VectorXd dot_products = M * vector_1;

        VectorXd result = dot_products.array() / (norm1 * all_norms.array());

        return result;
    }

    static double n_similarity(const std::vector<VectorXd>& vector_list_1, const std::vector<VectorXd>& vector_list_2) {
        if (vector_list_1.empty() || vector_list_2.empty()) {
            throw std::invalid_argument("At least one of the passed list is empty.");
        }

        Index d1 = vector_list_1[0].size();
        for (const auto& v : vector_list_1) {
            if (v.size() != d1) {
                throw std::invalid_argument("Vectors in vector_list_1 must have the same dimension.");
            }
        }
        Index d2 = vector_list_2[0].size();
        for (const auto& v : vector_list_2) {
            if (v.size() != d2) {
                throw std::invalid_argument("Vectors in vector_list_2 must have the same dimension.");
            }
        }
        if (d1 != d2) {
            throw std::invalid_argument("The two sets of vectors must have the same dimension.");
        }

        VectorXd mean1 = VectorXd::Zero(d1);
        for (const auto& v : vector_list_1) {
            mean1 += v;
        }
        mean1 /= vector_list_1.size();

        VectorXd mean2 = VectorXd::Zero(d2);
        for (const auto& v : vector_list_2) {
            mean2 += v;
        }
        mean2 /= vector_list_2.size();

        return similarity(mean1, mean2);
    }

    static std::map<std::string, double> compute_idf_weight_dict(int total_num, const std::map<std::string, double>& number_dict) {
        std::map<std::string, double> result;
        for (const auto& kv : number_dict) {
            result[kv.first] = std::log((total_num + 1) / (kv.second + 1));
        }
        return result;
    }

private:
    static VectorXd unitvec(const VectorXd& vec) {
        double norm_val = vec.norm();
        if (norm_val == 0.0) {
            norm_val = 1.0;
        }
        return vec / norm_val;
    }
};

File: data/derived_cot/rq3_prompting/results/translation/qwen/cpp_to_py/code_output/VectorUtil.py
import math
from typing import List, Dict

def norm(vec: List[float]) -> float:
    if not vec:
        return 0.0
    return math.sqrt(sum(x * x for x in vec))

def normalize(vec: List[float]) -> List[float]:
    norm_val = norm(vec)
    if norm_val == 0.0:
        return []
    return [x / norm_val for x in vec]

def similarity(vec1: List[float], vec2: List[float]) -> float:
    n1 = normalize(vec1)
    n2 = normalize(vec2)
    if len(n1) != len(n2):
        raise ValueError("Vectors must have the same length")
    return sum(x * y for x, y in zip(n1, n2))

def cosine_similarities(vec: List[float], vec_list: List[List[float]]) -> List[float]:
    if not vec_list:
        return []
    n = len(vec)
    for v in vec_list:
        if len(v) != n:
            raise ValueError("All vectors must have the same length as the given vector")
    norm_vec1 = norm(vec)
    result = []
    for v in vec_list:
        norm_vec_all = norm(v)
        if norm_vec_all == 0.0:
            result.append(0.0)
            continue
        dot_product = sum(x * y for x, y in zip(vec, v))
        similarity = dot_product / (norm_vec1 * norm_vec_all)
        result.append(similarity)
    return result

def n_similarity(vec_list1: List[List[float]], vec_list2: List[List[float]]) -> float:
    if not vec_list1 or not vec_list2:
        raise ValueError("At least one of the lists is empty")
    n = len(vec_list1[0]) if vec_list1 else 0
    for vec in vec_list1:
        if len(vec) != n:
            raise ValueError("All vectors in vec_list1 must have the same length")
    n2 = len(vec_list2[0]) if vec_list2 else 0
    for vec in vec_list2:
        if len(vec) != n2:
            raise ValueError("All vectors in vec_list2 must have the same length")
    if n != n2:
        raise ValueError("Vectors in vec_list1 and vec_list2 must have the same length")
    mean1 = [0.0] * n
    mean2 = [0.0] * n
    for vec in vec_list1:
        for i, x in enumerate(vec):
            mean1[i] += x
    for vec in vec_list2:
        for i, x in enumerate(vec):
            mean2[i] += x
    mean1 = [x / len(vec_list1) for x in mean1]
    mean2 = [x / len(vec_list2) for x in mean2]
    return similarity(mean1, mean2)

def compute_idf_weight_dict(total_num: int, number_dict: Dict[str, float]) -> Dict[str, float]:
    if not number_dict:
        return {}
    keys = list(number_dict.keys())
    counts = [number_dict[key] for key in keys]
    idf_weights = [math.log((total_num + 1) / (count + 1)) for count in counts]
    return dict(zip(keys, idf_weights))